cmd/snap/cmd_info.go

```diff
@@ -163,27 +163,68 @@ func coalesce(snaps ...*client.Snap) *client.Snap {
 	return nil
 }
 
-// formatDescr formats a given string (typically a snap description)
+func rTrimRightSpace(text []rune) []rune {
+	j := len(text)
+	for j > 0 && unicode.IsSpace(text[j-1]) {
+		j--
+	}
+	return text[:j]
+}
+
+func rLastIndexSpace(a []rune) int {
+	for i := len(a) - 1; i >= 0; i-- {
+		if unicode.IsSpace(a[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// this is _wrong_ for much of unicode (because the width of a rune on the
+// terminal is anything between 0 and 2, not always 1 as this code assumes)
+// but fixing that is Hard. Long story short, you can get close using a couple
+// of big unicode table. Getting it 100% requires a terminfo-alike of unicode
+// behaviour.
+// This (and probably formatDescr below) should move to strutil once we're
+// happy with it getting wider use.
+func wrap1(out io.Writer, text []rune, width int) {
```

Comment:
`s/text/line`, or maybe rename to `wrapLine`, calling it with freefrom text will produce undesired results.

I think you could easily take a `text string` if you drop direct increment in loops and instead use range. For instance
```
func wrap1(..., text string, ..) {
    var idx int
    var c rune
    for idx, c = range text {
        if unicode.IsSpace(c) { break }
    }
    dent, text = " " + text[:idx]
    text = text[idx:]
    ...
}
```

Reply:
if you can write it as something taking a `string`, that doesn't end up going to and from `[]rune` (or doing `utf8.RuneCountInString`) anyway, teach me :-)

Reply:
Best timeboxed version I could come up with (that also works for multibyte characters):
```
func wrap1(out io.Writer, text string, width int) {
	text = strings.TrimRightFunc(text, unicode.IsSpace)
	var idx int
	var c rune
	for idx, c = range text {
		if !unicode.IsSpace(c) {
			break
		}
	}
	dent := "  " + string(text[:idx])
	output := func(what string) {
		fmt.Fprint(out, dent)
		fmt.Fprintln(out, what)
	}
	text = text[idx:]
	width -= idx + 2
	for len([]rune(text)) > width {
		idx = len(text)
		for idx > 0 && len([]rune(text[:idx])) > width {
			idx = strings.LastIndexFunc(text[:idx], unicode.IsSpace)
		}
		if idx < 0 {
			idx = width
		}
		output(text[:idx])
		text = text[idx:]

		for idx, c = range text {
			if !unicode.IsSpace(c) {
				break
			}
			idx++
		}
		text = text[idx:]
	}

	output(text)
}
```
Maybe a stream oriented version would be nicer. Anyways, I see what you mean now.

Reply:
this works: https://pastebin.ubuntu.com/p/yPNh9YB2xQ/
is it better?

Reply:
if it's any consolation, in go 1.10 the naive version is faster (still uses more memory though):
1: []rune, strings.Split; 2: []rune, IndexByte to avoid building the slice; 3: string, strings.Split; 4: string, IndexByte.
Go 1.6:
```
Benchmark1-8	     300	   5854822 ns/op	  47.47 MB/s	 1686576 B/op	   25339 allocs/op
Benchmark2-8	     200	   6317663 ns/op	  44.00 MB/s	 1751898 B/op	   26588 allocs/op
Benchmark3-8	     300	   4055336 ns/op	  68.54 MB/s	  267723 B/op	   16914 allocs/op
Benchmark4-8	     300	   4530021 ns/op	  61.36 MB/s	  333045 B/op	   18163 allocs/op
```
Go 1.10:
```
Benchmark1-8   	     500	   3642264 ns/op	  76.31 MB/s	 1587784 B/op	   19264 allocs/op
Benchmark2-8   	     500	   3783451 ns/op	  73.46 MB/s	 1653233 B/op	   20513 allocs/op
Benchmark3-8   	     500	   3665364 ns/op	  75.83 MB/s	  170418 B/op	   10837 allocs/op
Benchmark4-8   	     500	   3742226 ns/op	  74.27 MB/s	  235858 B/op	   12086 allocs/op
```

Reply:
Nice, thanks for benchmarking this!

Reply:
Haven't looked at the two versions yet, but let's please focus on clarity and correctness instead of performance. This logic is not speed or allocation sensitive since it's very rarely used in computing terms, and when it's used it's used very few times (one?).

Reply:
@niemeyer that was my point, yes; it was a fun exercise for me to do on my day off, but I wouldn't propose it for this function of my own volition.

In a library that deals with these issues directly and exclusively (like the one I'm slowly writing in my free time), sure :-)